Deletion: rework list_and_mark_unlocked_replicas. Closes #4897

rcarpa opened the pull request:
Rework the workflow for getting deletion replicas. Refer to individual commits for details.

Commit:
Deletion: use temporary tables to list replicas to delete. Closes #4897

Due to reasons described in detail in the linked ticket, I didn't
find any way to integrate the "limit" argument into the initial
query. We are forced to "over-shoot" and get more data from the
database. Some of these retrieved rows will be skipped latter
because they don't match additional conditions required for safe
removal (no sources; no pending transfers for last replicas).

The workflow is however completely re-worked:
- We still have an unbound query which is iterated in bulks via
`yield_per`, but this iteration now depends on the limit argument.
- We use a temporary table on the database side to help with
filtering bulks rather than performing a separate query for each
request.

Commit:
Deletion: homogenize if/else in list_and_mark_unlocked_replicas. #4897

The commit changes the behavior of the function, but it seems more
correct this way.

First, it doesn't seem logical to check the obsolete condition inside
`if state != ReplicaState.UNAVAILABLE'. Moreover, this is only done for
the case when a file has more than one replica, which is even more
confusing. It looks more correct to move the check outside the if,
so that tombstone is also checked for unavailable replicas.

Second, it's true that we don't want to count the unavailable replicas
size into total_bytes. These replicas don't exist on disk, so deleting
them will not free any space. But why isn't it done for the branch in
which we analyze the last existing replica? Change the code to have
the same behavior when a single replica exists.

Third, I don't understand why we don't count the size of obsolete
replicas into total_bytes. Move total_bytes incrementation outside
`if tombstone != OBSOLETE`.

Fourth, seems wrong to not count unavailable replicas in the
"total_files". According to function definition, the "limit" argument
defines how many replicas should be returned by the function.

bari12 commented:
- I agree with the changes in the first commit. I am not entirely sure why it was done differently before. @cserf ?
- The second commit looks ok, but the whole `(2 * limit), math.ceil(1.25 * limit)` doesn't feel very good. I don't have a better idea and perhaps this is just not doable due to the reasons you explained in the ticket. I think it's still an improvement though.
@cserf ?

cserf commented on lib/rucio/core/replica.py:
```diff
@@ -2018,6 +2022,146 @@ def list_and_mark_unlocked_replicas(limit, bytes_=None, rse_id=None, delay_secon
+             models.RSEFileAssociation.name == models.Source.name,
+             models.RSEFileAssociation.rse_id == models.Source.rse_id)
+    ).where(
+        models.Source.scope.is_(None)  # Only try to delete replicas if they are not used as sources in any transfers
```
This is what we had in the past, but had to relax it because it can completely block the deletion of some files, e.g. : you have a file on RSE A and RSE B and a request to transfer to RSE C that is unavailable. In that case there is a row in the source table for RSE A and B, and there is no way to delete at least one of these copies until RSE C is back online.

rcarpa replied:
I'm not sure what you mean by "relax". To my understanding this is just literal translation of the current behavior : https://github.com/rucio/rucio/blob/e7f1f44e00f44349a146b8e9e93ab942996896da/lib/rucio/core/replica.py#L2044 

Did I make a mistake and changed the current behavior in some way I didn't realize?

bari12 replied:
This check was re-introduced by us last year, we did have to relax it before, that's what Cedric mentions.
But I think it is fine now, this PR does not change that behaviour.

cserf commented on lib/rucio/core/replica.py:
```diff
@@ -2018,6 +2022,146 @@ def list_and_mark_unlocked_replicas(limit, bytes_=None, rse_id=None, delay_secon
+        models.RSEFileAssociation.tombstone
+    ).with_for_update(
+        skip_locked=True,
+        # oracle: we must specify a column, not a table; however, it doesn't matter which column, the lock is put on the whole row
```
Oh. Didn't know about this. Good to know

cserf commented on lib/rucio/core/replica.py:
```diff
@@ -2018,6 +2022,146 @@ def list_and_mark_unlocked_replicas(limit, bytes_=None, rse_id=None, delay_secon
+        of=models.RSEFileAssociation.scope,
+    )
+
+    for chunk in chunks(session.execute(stmt).yield_per(2 * limit), math.ceil(1.25 * limit)):
```
Magic numbers ;) ?

rcarpa replied:
Yes, but probably not any more magic than the 1000 in the existing code: 
https://github.com/rucio/rucio/blob/e7f1f44e00f44349a146b8e9e93ab942996896da/lib/rucio/core/replica.py#L2055

cserf reviewed (commented):
The logic looks good, now we need to evaluate at scale

bari12 commented:
Please rebase, unless there are any more comments, it's gtg

Commit:
Deletion: rewrite setting BEING_DELETED without or-chunk

rcarpa commented:
I rebased; but I also added one additional commit. I thought this change was already included in this PR, but I only pushed it to one of my personal branches. Please review the last commit too.
